README.md
An ssl monitor for nginx
========================

SSLMON is a SSL MONitor module for nginx, providing statistical information about ssl.
It's designed to be used in a ssl terminator configuration, but it can be used also
wihtout upstream. In this case, the upstream time is always 0. 
Tested with nginx 1.10.3 and 1.12.0

How to build
------------
There are four patch files:

- patch-auto-modules
- patch-auto-options
- patch-src-event-ngx_event_openssl.c
- patch-src-http-ngx_http_request.c

to be applied to nginx sources; the patches are optional and provide handshake error statistics.
The patc

Then, you can configure nginx:

```bash
./configure --prefix=/opt/nginx --add-module=/path/to/sslmon --with-http_sslmon_module
```

The `--with-https_sslmon_module` option is provided by the patches.

Then, an easy

```bash
make && make install
```

will do the rest.

Preparation
-----------
The sslmon module write statistics in one file per worker:

- `/var/log/nginx/sslmon/sslmon-12345.log`
- `/var/log/nginx/sslmon/sslmon-12346.log`

where the number is the ProcessID of the worker.
The `/var/log/nginx/sslmon` directory has to exists and writeble from the user owner of the worker
(on FreeBSD systems is the `www` user, in other can be the `nginx` user)

Unfortunately, the path is currently hardcoded.

Configuration parameters
------------------------
The sslmon provides two parameters to customize the behavior:

```
sslmon_slow_request_time 100;
sslmon_update_period 15;
```

The sslmon module counts the number of slow request. A request is slow if it's response time
is greater then a specific threshold; this threshold can ne specified via the `sslmon_slow_request_time` parameter. The default value is 200 ms.

The sslmon rewrite the statistic file every 5 minutes. If you want slower or faster update, you can tune this period with the parameter `sslmon_update_period`, in seconds. The default value is 300 seconds (5 minutes)

Error Log
---------
The sslmon module can write to the error log.
Errors are logged at NGX_LOG_ERROR
Notice level provide only some status information.
Debug level can be really verbose.

Statistic provided
------------------
The collected statistics are:

request counter
slow request counter
ssl errors counter (handshake failures)
reused sessions counter
response time (avg)
upstream time (avg)
ssl time (avg) [response-upstream]

In the statistic log file you have also both configuration parameters and the timestamp expressed as epoch time.

An additional statistic information is the cipher type. It's a counters list of all matched ciphers used in the period. The list shows only the ciphers used in the last period, it's not a fixed list.

Collectd support
----------------
In the collectd directory, a shell script is provided to work with collectd.
The script reads all the statistic files and print them out (summed up) in a collectd compliant format.
The script can be used as exec plugin in collectd.
The script doesn't support the chiper counters




CHANGELOG.md
# Nginx SSLMon(itor) module

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).
## [Unreleased]

## [1.1.1] - 2017-07-07
### Fixed
- Upstream time happens to be a big number sometime. Hopefully fixed.

### Added
- Capping the upstream time using the response time as max limit.

## [1.1.0] - 2017-05-03
### Added
- Support to nginx v1.12.x
- Added folders v1.12.x and v1.10.x for version specific stuff
- Added the 3-clause BSD license

### Changed
- Start adopting the Semanting Versioning
- README.md is properly using the markdown




collectd/merger.awk
#!/usr/bin/env -S awk -f
BEGIN { epoch_counter=0;
	epoch=0;
	counter=0;
	slow_requests=0;
	reused_sessions=0;
	avg_rt=0.0;
	avg_ut=0.0;
	avg_net_rt=0.0;
	FS="=";
}

/^epoch/	{ if ( epoch < $2 )
			epoch = $2;
		  epoch_counter++;
		}
/^counter/		{ counter = counter + $2 }
/^slow_requests/	{ slow_requests = slow_requests + $2 }
/^reused_sessions/	{ reused_sessions = reused_sessions + $2 }
/^avg_rt/		{ avg_rt = avg_rt + $2 }
/^avg_ut/		{ avg_ut = avg_ut + $2 }
/^avg_net_rt/		{ avg_net_rt = avg_net_rt + $2 }

END {
	if ( epoch_counter == 0 )
		epoch_counter = 1;
	avg_rt = avg_rt / epoch_counter;
	avg_ut = avg_ut / epoch_counter;
	avg_net_rt = avg_net_rt / epoch_counter;
	printf "epoch=%d\n",epoch
	printf "counter=%d\n",counter
	printf "slow_requests=%d\n", slow_requests
	printf "reused_sessions=%d\n", reused_sessions
	printf "avg_rt=%d\n", avg_rt
	printf "avg_ut=%d\n", avg_ut
	printf "avg_net_rt=%d\n", avg_net_rt
}


collectd/sslmon-collectd.sh
#!/bin/sh
UPDATE_PERIOD=15
IDENTIFIER="$(hostname)"/ssl/

merger()
{
	awk '
	BEGIN { epoch_counter=0;
		epoch=0;
		counter=0;
		slow_requests=0;
		ssl_errors=0;
		reused_sessions=0;
		avg_rt=0.0;
		avg_ut=0.0;
		avg_net_rt=0.0;
		FS="=";
	}
	/^epoch/	{ if ( epoch < $2 )
				epoch = $2;
			  epoch_counter++;
			}
	/^counter/		{ counter = counter + $2 }
	/^slow_requests/	{ slow_requests = slow_requests + $2 }
	/^ssl_errors/		{ ssl_errors = ssl_errors + $2 }
	/^reused_sessions/	{ reused_sessions = reused_sessions + $2 }
	/^avg_rt/		{ avg_rt = avg_rt + $2 }
	/^avg_ut/		{ avg_ut = avg_ut + $2 }
	/^avg_net_rt/		{ avg_net_rt = avg_net_rt + $2 }
	END {
		if ( epoch_counter == 0 )
			epoch_counter = 1;
		avg_rt = avg_rt / epoch_counter;
		avg_ut = avg_ut / epoch_counter;
		avg_net_rt = avg_net_rt / epoch_counter;
		printf "epoch=%d\n",epoch;
		printf "counter=%d\n",counter;
		printf "slow_requests=%d\n", slow_requests;
		printf "ssl_errors=%d\n", ssl_errors;
		printf "reused_sessions=%d\n", reused_sessions;
		printf "avg_rt=%d\n", avg_rt;
		printf "avg_ut=%d\n", avg_ut;
		printf "avg_net_rt=%d\n", avg_net_rt;
	}' "$@"
}

while true;
do
	start_epoch=$(date +%s)
	if [ -z "$(ls /var/log/nginx/sslmon/sslmon-* 2>/dev/null)" ]; then
		data_epoch=${start_epoch}
		requests=0
		slow_requests=0
		ssl_errors=0
		reused_sessions=0
		request_time=0
		upstream_time=0
		ssl_time=0
	else
		MERGEFILE=$(mktemp -t sslmon-XXX)
		merger /var/log/nginx/sslmon/* > ${MERGEFILE}
		data_epoch=$(awk -F= '/^epoc/ { print $2 }' ${MERGEFILE})
		if [ ${data_epoch} -eq 0 ]; then
			data_epoch=${start_epoch}
		fi
		requests=$(awk -F= '/^counter/ { print $2 }' ${MERGEFILE})
		slow_requests=$(awk -F= '/^slow_requests/ { print $2 }' ${MERGEFILE})
		ssl_errors=$(awk -F= '/^ssl_errors/ { print $2 }' ${MERGEFILE})
		reused_sessions=$(awk -F= '/^reused_sessions/ { print $2 }' ${MERGEFILE})
		request_time=$(awk -F= '/^avg_rt/ { print $2 }' ${MERGEFILE})
		upstream_time=$(awk -F= '/^avg_ut/ { print $2 }' ${MERGEFILE})
		ssl_time=$(awk -F= '/^avg_net_rt/ { print $2 }' ${MERGEFILE})
#		cat ${MERGEFILE}
		rm -rf ${MERGEFILE}
	fi
	#rrdtool updatev sslmon.rrd -s ${data_epoch}:${requests}:${slow_requests}:${reused_sessions}:${request_time}:${upstream_time}:${ssl_time}
	echo PUTVAL "${IDENTIFIER}counter-requests" interval=$UPDATE_PERIOD ${data_epoch}:${requests}
	echo PUTVAL "${IDENTIFIER}counter-slow_requests" interval=$UPDATE_PERIOD ${data_epoch}:${slow_requests}
	echo PUTVAL "${IDENTIFIER}counter-ssl_errors" interval=$UPDATE_PERIOD ${data_epoch}:${ssl_errors}
	echo PUTVAL "${IDENTIFIER}counter-reused_sessions" interval=$UPDATE_PERIOD ${data_epoch}:${reused_sessions}
	echo PUTVAL "${IDENTIFIER}total_time_in_ms-request_time" interval=$UPDATE_PERIOD ${data_epoch}:${request_time}
	echo PUTVAL "${IDENTIFIER}total_time_in_ms-upstream_time" interval=$UPDATE_PERIOD ${data_epoch}:${upstream_time}
	echo PUTVAL "${IDENTIFIER}total_time_in_ms-ssl_time" interval=$UPDATE_PERIOD ${data_epoch}:${ssl_time}
	now=$(date +%s)
	if [ $(( now - start_epoch )) -le $UPDATE_PERIOD ]; then
		sleep $(( $UPDATE_PERIOD + start_epoch - now ))
	fi
done


patch-src-event-ngx_event_openssl.c
--- src/event/ngx_event_openssl.c.orig	2016-11-23 13:46:35.524339000 +0100
+++ src/event/ngx_event_openssl.c	2016-11-23 13:53:46.309927000 +0100
@@ -17,6 +17,10 @@
     ngx_uint_t  engine;   /* unsigned  engine:1; */
 } ngx_openssl_conf_t;
 
+#if (NGX_HTTP_SSLMON)
+extern void
+ngx_http_sslmon_increment_ssl_handshake_errors();
+#endif
 
 static int ngx_ssl_password_callback(char *buf, int size, int rwflag,
     void *userdata);
@@ -1214,10 +1218,18 @@
         c->write->handler = ngx_ssl_handshake_handler;
 
         if (ngx_handle_read_event(c->read, 0) != NGX_OK) {
+#if (NGX_HTTP_SSLMON)
+            /* counting handshake errors */
+            ngx_http_sslmon_increment_ssl_handshake_errors();
+#endif
             return NGX_ERROR;
         }
 
         if (ngx_handle_write_event(c->write, 0) != NGX_OK) {
+#if (NGX_HTTP_SSLMON)
+            /* counting handshake errors */
+            ngx_http_sslmon_increment_ssl_handshake_errors();
+#endif
             return NGX_ERROR;
         }
 
@@ -1230,10 +1242,18 @@
         c->write->handler = ngx_ssl_handshake_handler;
 
         if (ngx_handle_read_event(c->read, 0) != NGX_OK) {
+#if (NGX_HTTP_SSLMON)
+            /* counting handshake errors */
+            ngx_http_sslmon_increment_ssl_handshake_errors();
+#endif
             return NGX_ERROR;
         }
 
         if (ngx_handle_write_event(c->write, 0) != NGX_OK) {
+#if (NGX_HTTP_SSLMON)
+            /* counting handshake errors */
+            ngx_http_sslmon_increment_ssl_handshake_errors();
+#endif
             return NGX_ERROR;
         }
 
@@ -1249,6 +1269,10 @@
     if (sslerr == SSL_ERROR_ZERO_RETURN || ERR_peek_error() == 0) {
         ngx_connection_error(c, err,
                              "peer closed connection in SSL handshake");
+#if (NGX_HTTP_SSLMON)
+        /* counting handshake errors */
+        ngx_http_sslmon_increment_ssl_handshake_errors();
+#endif
 
         return NGX_ERROR;
     }
@@ -1256,6 +1280,10 @@
     c->read->error = 1;
 
     ngx_ssl_connection_error(c, sslerr, err, "SSL_do_handshake() failed");
+#if (NGX_HTTP_SSLMON)
+    /* counting handshake errors */
+    ngx_http_sslmon_increment_ssl_handshake_errors();
+#endif
 
     return NGX_ERROR;
 }


ngx_http_sslmon.c
/* Put a license-something here */

#include <ngx_config.h>
#include <ngx_core.h>
#include <ngx_http.h>
#include <nginx.h>

/* ************************ */
/* Types/structs definition */
/* ************************ */

typedef long unsigned ngx_cnt;	/* counter */

typedef struct {
	char * name;
	ngx_cnt counter;
} ngx_http_sslmon_ciphers_cnt_t;

typedef struct {
	ngx_cnt counter;		/* requests counter */
	ngx_cnt rt_sum;			/* req time sum */
	ngx_cnt ut_sum;			/* upstream time sum */
	ngx_cnt slow_requests;		/* requests slower than 200ms */
	ngx_cnt reused_sessions;	/* reused sessions */
	ngx_array_t* ciphers_cnt;	/* ciphers counter */
} ngx_http_sslmon_stats_t;

typedef struct {
	ngx_uint_t			update_period;
	ngx_uint_t			slow_request_time;
	int 				fd;
	ngx_http_sslmon_stats_t * 	stats;
	ngx_str_t			filename;
} ngx_http_sslmon_main_conf_t;

/* ****************************** */
/* Function prototype definitions */
/* ****************************** */

void
ngx_http_sslmon_increment_ssl_handshake_errors();

static ngx_int_t
ngx_http_sslmon_init( ngx_conf_t *cf );

static void *
ngx_http_sslmon_create_main_conf( ngx_conf_t * cf );

static char *
ngx_http_sslmon_merge_main_conf(ngx_conf_t *cf, void *conf);

static ngx_int_t
ngx_http_sslmon_init_process( ngx_cycle_t * cx );

static void
ngx_http_sslmon_exit_process( ngx_cycle_t * cx );

static void
ngx_http_sslmon_set_timer( ngx_http_sslmon_main_conf_t *c, ngx_log_t *l );

static void
ngx_http_sslmon_timer_handler( ngx_event_t *ev );

static void
ngx_http_sslmon_write_report( ngx_http_sslmon_main_conf_t *conf, ngx_log_t *l );

/* ********************** */
/* Global variables timer */
/* ********************** */

ngx_cnt ngx_http_sslmon_ssl_handshake_errors;	/* SSL handshake errors */
ngx_event_t ngx_http_sslmon_timer;		/* write report timer */

/* ************************ */
/* Nginx module definitions */
/* ************************ */

/* How often the statistic file has to be written */
/* default: every 300 seconds */
#define SSLMON_DEFAULT_UPDATE_PERIOD 300*1000

/* A reuqest is considered slow if it takes more than 200ms */
#define SSLMON_DEFAULT_SLOW_REQUEST_TIME 200
static ngx_command_t ngx_http_sslmon_commands[] = {

        { ngx_string("sslmon_update_period"),
          NGX_HTTP_MAIN_CONF | NGX_CONF_TAKE1,
          ngx_conf_set_msec_slot,
          NGX_HTTP_MAIN_CONF_OFFSET,
	  offsetof(ngx_http_sslmon_main_conf_t, update_period),
	  NULL },

        { ngx_string("sslmon_slow_request_time"),
          NGX_HTTP_MAIN_CONF | NGX_CONF_TAKE1,
          ngx_conf_set_num_slot,
          NGX_HTTP_MAIN_CONF_OFFSET,
	  offsetof(ngx_http_sslmon_main_conf_t, slow_request_time),
	  NULL },
	ngx_null_command
};

static ngx_http_module_t  ngx_http_sslmon_module_ctx = {
	NULL,					/* preconfiguration */
	ngx_http_sslmon_init,			/* postconfiguration */

	ngx_http_sslmon_create_main_conf,	/* create main configuration */
	ngx_http_sslmon_merge_main_conf,	/* init main configuration */

	NULL,					/* create server configuration */
	NULL,					/* merge server configuration */

	NULL,					/* create location configration */
	NULL,					/* merge location configration */
};

ngx_module_t ngx_http_sslmon_module = {
	NGX_MODULE_V1,
	&ngx_http_sslmon_module_ctx,	/* module context */
	ngx_http_sslmon_commands,	/* module directives */
	NGX_HTTP_MODULE,		/* module type */
	NULL,				/* init master */
	NULL,				/* init module */
	ngx_http_sslmon_init_process,	/* init process */
	NULL,				/* init thread */
	NULL,				/* exit thread */
	ngx_http_sslmon_exit_process,	/* exit process */
	NULL,				/* exit master */
	NGX_MODULE_V1_PADDING
};

static void
ngx_http_sslmon_add_cipher(ngx_array_t * cnt_array, const char * cipher_name)
{
	ngx_http_sslmon_ciphers_cnt_t cc, * ccp;
	cc.name = (char *)cipher_name;
	cc.counter = 1;
	ccp = ngx_array_push(cnt_array);
	if( ccp == NULL ) {
		/* disaster */
		return;
	}
	*ccp = cc;
}

static int
ngx_http_sslmon_find_and_incr_cipher( ngx_array_t * cnt_array, const char * cipher_name)
{
	ngx_uint_t i = 0;
	ngx_http_sslmon_ciphers_cnt_t * ciphers = cnt_array->elts;
	size_t len = strlen(cipher_name);
	for( i=0; i < cnt_array->nelts ; i++) {
		size_t len2 = strlen( ciphers[i].name );
		if( strncmp( cipher_name, ciphers[i].name, ngx_max(len,len2) ) == 0 ) {
			ciphers[i].counter++;
			return 1; /* true */
		}
	}
	return 0; /* false */
}

static void *
ngx_http_sslmon_create_main_conf(ngx_conf_t *cf)
{
	ngx_http_sslmon_main_conf_t *conf;
	conf = ngx_pcalloc(cf->pool, sizeof(ngx_http_sslmon_main_conf_t));
	if( conf == NULL ) {
		ngx_log_error(NGX_LOG_ERR, cf->log, 0,
			"sslmon_create_main_conf: not able to allocate");
		return NULL;
	} else {
		ngx_log_error(NGX_LOG_DEBUG, cf->log, 0,
			"sslmon_create_main_conf: conf store at %p", conf);
	}
	conf->update_period = NGX_CONF_UNSET_UINT;
	conf->slow_request_time = NGX_CONF_UNSET_UINT;
	conf->fd = NGX_CONF_UNSET;

	ngx_http_sslmon_stats_t *stats;
	stats = ngx_pcalloc(cf->pool, sizeof(ngx_http_sslmon_stats_t));
	if( stats == NULL ) {
		ngx_log_error(NGX_LOG_ERR, cf->log, 0,
			"sslmon_create_main_conf: not able to allocate stats");
		return NULL;
	} else {
		ngx_log_error(NGX_LOG_DEBUG, cf->log, 0,
			"sslmon_create_main_conf: stats store at %p", conf);
	}
	conf->stats = stats;
	stats->ciphers_cnt =
		ngx_array_create( cf->pool, 15, sizeof( ngx_http_sslmon_ciphers_cnt_t ));
	return conf;
}

void
ngx_http_sslmon_increment_ssl_handshake_errors()
{
	ngx_http_sslmon_ssl_handshake_errors++;
}

static void
ngx_http_sslmon_reset_stats( ngx_http_sslmon_stats_t * s )
{
	/* resetting counters */
	s->counter = s->slow_requests = s->reused_sessions = 0;
	s->rt_sum = s->ut_sum = 0;
	ngx_array_init( s->ciphers_cnt, s->ciphers_cnt->pool, 25,
		sizeof( ngx_http_sslmon_ciphers_cnt_t ));
	ngx_http_sslmon_ssl_handshake_errors = 0;
}

static char *
ngx_http_sslmon_merge_main_conf(ngx_conf_t *cf, void *c)
{
	ngx_http_sslmon_main_conf_t * conf = c;
	ngx_http_sslmon_stats_t * stats = conf->stats;
	if( conf->update_period == NGX_CONF_UNSET_UINT ) {
		conf->update_period = SSLMON_DEFAULT_UPDATE_PERIOD;
	}
	if( conf->slow_request_time == NGX_CONF_UNSET_UINT ) {
		conf->slow_request_time = SSLMON_DEFAULT_SLOW_REQUEST_TIME;
	}
	ngx_log_error(NGX_LOG_NOTICE, cf->log, 0,
		"sslmon_merge_main_conf: update_period %l s, slow_request %l ms",
		conf->update_period/1000, conf->slow_request_time);
	ngx_http_sslmon_reset_stats( stats );
	return NGX_CONF_OK;
}

static void
ngx_http_sslmon_set_timer( ngx_http_sslmon_main_conf_t *c, ngx_log_t *log )
{
	ngx_http_sslmon_timer.handler = ngx_http_sslmon_timer_handler;
	ngx_http_sslmon_timer.log = log;
	ngx_http_sslmon_timer.data = c;
	ngx_add_timer( &ngx_http_sslmon_timer, c->update_period );
}

#define SSLMON_FILENAME_BASE "/var/log/nginx/sslmon/sslmon"
#define SSLMON_FILENAME_MAXSIZE 512
static ngx_int_t
ngx_http_sslmon_init_process( ngx_cycle_t * cx )
{
	ngx_http_sslmon_main_conf_t * conf =
		ngx_http_cycle_get_module_main_conf( cx, ngx_http_sslmon_module );
	if( conf == NULL ) { /* paranoid check */
		ngx_log_error(NGX_LOG_ERR, cx->log, 0,
			"sslmon_init_process: not conf found");
		return NGX_OK;
	}

/* create the log filename appending the pid */
	ngx_pid_t pid = ngx_getpid();
	conf->filename.data = ngx_pcalloc(cx->pool, SSLMON_FILENAME_MAXSIZE);
	if( conf->filename.data == NULL ) {
		ngx_log_error(NGX_LOG_ERR, cx->log, 0,
			"sslmon_init_process:: not able to allocate");
		return NGX_ERROR;
	}
	conf->filename.len = 0;
	conf->filename.len = snprintf( (char *)conf->filename.data, SSLMON_FILENAME_MAXSIZE,
		"%s-%d.log", SSLMON_FILENAME_BASE,pid );
	ngx_log_error(NGX_LOG_NOTICE, cx->log, 0,
		"sslmon_init_process: sslmon log file %s", conf->filename.data );
/* opening the log file */
	conf->fd = ngx_open_file( conf->filename.data, NGX_FILE_WRONLY,
		NGX_FILE_TRUNCATE | NGX_FILE_NONBLOCK,
		S_IRUSR|S_IWUSR|S_IRGRP|S_IROTH );

	if ( conf->fd == -1 ) {
		ngx_log_error(NGX_LOG_ERR, cx->log, 0,
			"sslmon_init_process: file not open (%d)", errno);
		conf->fd = NGX_CONF_UNSET;
		return NGX_ERROR;
	}
/* preparing the timer */
	ngx_http_sslmon_set_timer( conf, cx->log );
	ngx_log_error(NGX_LOG_DEBUG, cx->log, 0,
		"sslmon_init_process: starting process %d - updating conf %p", pid, conf );
	return NGX_OK;
}

static void
ngx_http_sslmon_exit_process( ngx_cycle_t * cx )
{
	ngx_http_sslmon_main_conf_t * conf =
		ngx_http_cycle_get_module_main_conf( cx, ngx_http_sslmon_module );
	(void)close( conf->fd );
	(void)unlink( (char *)conf->filename.data );
}

static ngx_http_variable_value_t *
ngx_http_sslmon_getvar( ngx_http_request_t *r, const char * cstr )
{
	ngx_str_t var_name = ngx_string(cstr);
	var_name.len=strlen(cstr);
	ngx_uint_t key;
	ngx_http_variable_value_t * vv;
	key = ngx_hash_key( var_name.data, var_name.len );
	vv = ngx_http_get_variable( r, &var_name, key );
	return vv;
}

static ngx_int_t
ngx_http_sslmon_msec_getvar( ngx_http_request_t *r, const char * cstr )
{
	char tmpstr[10];
	bzero(tmpstr, 10);
	ngx_http_variable_value_t * vv;
	vv = ngx_http_sslmon_getvar( r, cstr );
	if( vv == NULL ) {
		ngx_log_error(NGX_LOG_ERR, r->connection->log, 0,
			"sslmon_handler: var %s is null", cstr);
		return 0;
	}
	if( vv->not_found ) {
		ngx_log_error(NGX_LOG_WARN, r->connection->log, 0,
			"sslmon_handler: no %s found", cstr);
		return 0;
	}
	int sec, msec;
	sec = msec = 0;
	strncpy( tmpstr, (char *)vv->data, vv->len );
	sscanf( tmpstr, "%d.%d", &sec, &msec );
	msec += 1000*sec;
	ngx_log_error(NGX_LOG_INFO, r->connection->log, 0,
		"sslmon_handler: %s %d", cstr, msec );
	return msec;
}

static ngx_int_t
ngx_http_sslmon_handler( ngx_http_request_t *r )
{
	ngx_http_sslmon_main_conf_t * conf;
	ngx_http_sslmon_stats_t * stats;
	unsigned int rt = 0; /* response time */
	unsigned int ut = 0; /* upstream time */
	unsigned long epoch = 0; /* request epoch */
	ngx_ssl_connection_t * ssl_connection;

	conf = ngx_http_get_module_main_conf( r, ngx_http_sslmon_module );
	stats = conf->stats;
	ssl_connection = r->connection->ssl;

	/* variable access is deprecated, we access directly the source of the information */
	/* rt = ngx_http_sslmon_msec_getvar( r, "request_time" ); */
	/* ut = ngx_http_sslmon_msec_getvar( r, "upstream_response_time" ); */

	/* get the response time */
	rt = (ngx_cached_time->sec - r->start_sec ) * 1000 + ngx_cached_time->msec - r->start_msec;
	ngx_log_error(NGX_LOG_DEBUG, r->connection->log, 0,
		"sslmon_handler: response time %d ms", rt);

	/* get the upstream time */
	if ( r->upstream_states == NULL || r->upstream_states->nelts == 0 ) {
		ut = 0;
	} else {
		ngx_uint_t i=0;
		ngx_msec_int_t ms = 0;
		ngx_http_upstream_state_t *state;
		state = r->upstream_states->elts;
		for( ;; ) {
			if( state[i].status ) {
				ms = state[i].response_time;
				if ( ms > 600000 ) {
					/* if the response time is bigger than 10 minutes */
					/* ms is holding an epoch in millisecond */
					 ngx_log_error(NGX_LOG_NOTICE, r->connection->log, 0,
						"sslmon_handler: state %d has huge ut", i);
					/* probably for a not yet completed request */
					ms = 0;
				}
			}
			ms = ngx_max(ms, 0);
			ut += ms;
			i++;
			if( i == r->upstream_states->nelts ) {
				break;
			}
			if (state[i].peer) {
				;
			} else {
				i++;
				if( i == r->upstream_states->nelts ) {
					break;
				}
				continue;
			}
		}
	}
	if ( ut > rt ) {
		ngx_log_error(NGX_LOG_ERR, r->connection->log, 0,
			"sslmon_handler: upstream time %d bigger than response time %d - Fixing",
			ut, rt);
		/* That shouldn't happen */
		ut = rt;
	}
	ngx_log_error(NGX_LOG_DEBUG, r->connection->log, 0,
		"sslmon_handler: upstream time %d ms", ut);
	if( rt > conf->slow_request_time ) {
		stats->slow_requests++;
	}
	stats->counter++;
	stats->rt_sum += rt;
	stats->ut_sum += ut;

	/* direct access to ssl information, avoiding variable parsing */
	if (ssl_connection && ssl_connection->connection ) {
		/* resued session counter */
		if( SSL_session_reused(ssl_connection->connection)) {
			stats->reused_sessions++;
			ngx_log_error(NGX_LOG_DEBUG, r->connection->log, 0,
				"sslmon_handler: ssl_session reused");
		}
		/* ciphers statistic */
		const char * cipher_name = SSL_get_cipher_name(ssl_connection->connection);
		if ( ngx_http_sslmon_find_and_incr_cipher( stats->ciphers_cnt , cipher_name) ) {
			ngx_log_error(NGX_LOG_DEBUG, r->connection->log, 0,
				"sslmon_handler: incrementing the counter %s", cipher_name);
		} else {
			ngx_http_sslmon_add_cipher( stats->ciphers_cnt, cipher_name );
			ngx_log_error(NGX_LOG_DEBUG, r->connection->log, 0,
				"sslmon_handler: adding a counter for the cipher %s", cipher_name);
		}
	}
	return NGX_OK;
}

static void
ngx_http_sslmon_write_report( ngx_http_sslmon_main_conf_t *conf, ngx_log_t *l )
{
	ngx_http_sslmon_stats_t * stats;
	stats = conf->stats;
	if( conf->fd != NGX_CONF_UNSET ) {
		int rc;
		ngx_log_error(NGX_LOG_DEBUG, l, 0,
			"sslmon_write_report: Updating report (counting %d) of pid %d",
			stats->counter, ngx_getpid());
		rc = ftruncate( conf->fd, 0 );
		if( rc != 0 ) {
			ngx_log_error(NGX_LOG_WARN, l, 0,
				"sslmon_write_report: ftruncate didn't work - %d", errno );
		}
		lseek( conf->fd, 0, SEEK_SET );
		dprintf( conf->fd, "update_period=%lu\n", conf->update_period/1000);
		dprintf( conf->fd, "slow_request_time=%lu\n", conf->slow_request_time);
		dprintf( conf->fd, "epoch=%lu\n", ngx_time() );
		dprintf( conf->fd, "counter=%lu\n", stats->counter );
		dprintf( conf->fd, "slow_requests=%lu\n", stats->slow_requests );
		dprintf( conf->fd, "reused_sessions=%lu\n", stats->reused_sessions );
		dprintf( conf->fd, "ssl_errors=%lu\n", ngx_http_sslmon_ssl_handshake_errors );
		if( stats->counter != 0 ) {
			dprintf( conf->fd, "avg_rt=%lf\n",
				stats->rt_sum/(double)(stats->counter) );
			dprintf( conf->fd, "avg_ut=%lf\n",
				stats->ut_sum/(double)(stats->counter) );
			dprintf( conf->fd, "avg_net_rt=%lf\n",
				(stats->rt_sum-stats->ut_sum)/(double)(stats->counter) );
		} else {
			dprintf( conf->fd, "avg_rt=0.0\n");
			dprintf( conf->fd, "avg_ut=0.0\n");
			dprintf( conf->fd, "avg_net_rt=0.0\n");
		}
		ngx_uint_t i = 0;
		ngx_http_sslmon_ciphers_cnt_t * ciphers = stats->ciphers_cnt->elts;
		for( i=0; i<stats->ciphers_cnt->nelts; i++ ) {
			if( ciphers[i].counter != 0 ) {
				dprintf( conf->fd, "cipher.%s=%lu\n",
					ciphers[i].name, ciphers[i].counter );
				ciphers[i].counter = 0;
			}
		}
	} else {
		ngx_log_error(NGX_LOG_ERR, l, 0,
			"sslmon_handler: fd not set - conf %p", conf);
	}
	/* reset all counters */
	ngx_http_sslmon_reset_stats( stats );
}

static void
ngx_http_sslmon_timer_handler( ngx_event_t *ev )
{
	ngx_http_sslmon_main_conf_t * conf;
	conf = ev->data;
	if ( ngx_exiting == 1 ) {
		ngx_log_error(NGX_LOG_NOTICE, ev->log, 0,
			"sslmon_timer_handler: quitting -> no timer anymore", conf);
	} else {
		ngx_http_sslmon_set_timer( conf, ev->log );
		ngx_log_error(NGX_LOG_DEBUG, ev->log, 0,
			"sslmon_timer_handler: re-setting timer", conf);
	}
	ngx_http_sslmon_write_report( conf, ev->log );
}

static ngx_int_t
ngx_http_sslmon_init( ngx_conf_t *cf )
{
	ngx_http_core_main_conf_t *cmcf;
	ngx_http_handler_pt *h;

	cmcf = ngx_http_conf_get_module_main_conf( cf, ngx_http_core_module );

	h = ngx_array_push(&cmcf->phases[NGX_HTTP_LOG_PHASE].handlers);
	if ( h == NULL ) {
		return NGX_ERROR;
	}

	*h = ngx_http_sslmon_handler;

	return NGX_OK;
}
